### PR TITLE
Modify fsfreeze behavior with paused VMs

### PR DESCRIPTION
**Pull request by alromeros**
**What this PR does / why we need it**:

This Pull Request aims to modify `fsfreeze` behavior so, in case of using it with a paused VM, it just returns successfully instead of triggering an error.

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: cmd/virt-freezer/main.go [hunk 1]
```diff
@@ -90,12 +103,24 @@ func main() {
...
+
 		err = client.FreezeVirtualMachine(vmi, *unfreezeTimeoutSeconds)
 		if err != nil {
 			log.Log.Reason(err).Error("Freezeing VMI failed")
 			os.Exit(1)
 		}
 	} else {
+		// Unfreeze is not possible when VM is paused
+		if hasPausedCondition(client) {
+			log.Log.Info("VM Paused: Unreeze is not possible")
```
Comment (brybacki) [inline cmd/virt-freezer/main.go:120]: Unreeze/Unfreeze

**Review by brybacki — APPROVED** [review]
The code is good.
 
There is one thing that concerns me.
- VM is paused - we try to freeze and it is a NO-OP,
- Some external action does Unfreeze, 
- We do backup or other actions assuming fs is frozen or vm is paused, but it is not. 

But I assume that is a general problem with the kubevir, as we cannot lock a VM for our own use, so it is probably not a problem here.

**Comment by brybacki** [comment]
@mhenriks what do you think about this PR?

**Review by awels — COMMENTED** [review]
Don't see any logical issues with the PR, not knowing that much about the code would it make sense to before you get to
```go
  if *freeze {
  ...
```
You simply do this:
```go
  if !hasPausedCondition(client) {
    if *freeze {
    ...
    } else {
    ...
    }
  }
  log.Log.Info("Exiting...")
```
And you simply skip the freeze/unfreeze if the VM is paused. The log stating the VM is paused might be nice, not sure. Logically this is the same, just fewer lines of code and no extra `os.Exit(0)`

File: cmd/virt-freezer/main.go [hunk 2]
```diff
@@ -42,6 +44,17 @@ func getGrpcClient() (cmdclient.LauncherClient, error) {
...
 
+// hasPausedCondition checks if the VM to be freezed/unfreezed is paused
+func hasPausedCondition(client cmdclient.LauncherClient) bool {
+	domain, exists, err := client.GetDomain()
+	if err != nil {
+		log.Log.Reason(err).Error("Failed to get domain")
+		os.Exit(1)
+	}
+
+	return exists && domain.Status.Status == api.Paused
```
Comment (mhenriks) [inline cmd/virt-freezer/main.go:55]: rather than checking paused, maybe it would be better to only freeze/unfreeze when status is `Running`?

**Review by mhenriks — COMMENTED** [review]
I kind of wonder if returning an error is actually the right thing to do.  I suspect it is possible that a paused vm has unflushed filesystem buffers in memory.  And this is a different from having no guest agent at all where a user presumably understands that there are no guarantees.  @brybacki what do you think?

File: cmd/virt-freezer/main.go (hunk 2 above)
Comment (alromeros) [reply cmd/virt-freezer/main.go:55]: I'm not too well-versed on what'd happen if we freeze in the other statuses, but it makes sense. I'll change it :+1:

**Comment by awels** [comment]
When you say returning an error is where are we returning the error? in the freeze/unfreeze if?

**Comment by mhenriks** [comment]
> When you say returning an error is where are we returning the error? in the freeze/unfreeze if?

virt-freezer returning non zero

**Comment by brybacki** [comment]
> 

Good question. And what do we do on VMExport? Is it allowed to export Paused VM?

**Comment by alromeros** [comment]
> > I kind of wonder if returning an error is actually the right thing to do. I suspect it is possible that a paused vm has unflushed filesystem buffers in memory. And this is a different from having no guest agent at all where a user presumably understands that there are no guarantees. @brybacki what do you think?
> 
> Good question. And what do we do on VMExport? Is it allowed to export Paused VM?

I think the export is only doable in stopped VMs. In running/paused VMs, the vmexport object just remains in pending phase until the VM is stopped.

**Pull request closed without merging**